src/commands/edit.ts

```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
  *
  * @keys `&` (kakoune: normal)
  */
-export function align(
-  _: Context,
-  selections: readonly vscode.Selection[],
+export function align(_: Context, fill: Argument<string> = " ") {
+  return edit((builder, selections) => {
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
+    const selectionByColumn: vscode.Selection[][] = [];
+    let currentLine: number | undefined = undefined;
+    let currentColumn: number = 0;
+    for (const selection of sortedSelections) {
+      if (selection.start.line !== selection.end.line) {
+        throw new Error(
+          "Cannot align selections that are spanning multiple lines",
+        );
+      }
 
-  fill: Argument<string> = " ",
-) {
-  const startChar = selections.reduce(
-    (max, sel) => (sel.start.character > max ? sel.start.character : max),
-    0,
-  );
+      if (selection.start.line !== currentLine) {
+        currentLine = selection.start.line;
+        currentColumn = 0;
+      }
 
-  return edit((builder, selections) => {
-    for (let i = 0, len = selections.length; i < len; i++) {
-      const selection = selections[i];
+      if (!(currentColumn in selectionByColumn)) {
+        selectionByColumn[currentColumn] = [];
+      }
```

Comment:
IIUC, `currentColumn` always increases by one, so this check can be replaced by:

```suggestion
      if (currentColumn === selectionsByColumn.length) {
        selectionsByColumn.push([]);
      }
```